metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -443,34 +444,90 @@ void convert_batch_to_row_diff(const std::string &pred_succ_fprefix,
 
     std::vector<std::unique_ptr<RowDiffColumnAnnotator>> row_diff(label_encoders.size());
 
+    const uint32_t files_open_per_thread
+            = MAX_NUM_FILES_OPEN / std::max((uint32_t)1, num_threads);
+    if (files_open_per_thread < 3) {
+        logger->error("Can't merge with less than 3 files open per thread");
+        exit(1);
+    }
+
     logger->trace("Generating row_diff columns...");
     #pragma omp parallel for num_threads(num_threads) schedule(dynamic)
     for (uint32_t l_idx = 0; l_idx < label_encoders.size(); ++l_idx) {
         std::vector<std::unique_ptr<bit_vector>> columns(label_encoders[l_idx].size());
 
         for (size_t j = 0; j < label_encoders[l_idx].size(); ++j) {
-            auto call_ones = [&](const std::function<void(uint64_t)>& call) {
+            auto call_ones = [&](const std::function<void(uint64_t)> &call) {
                 std::vector<std::string> filenames;
                 for (uint32_t chunk = 0; chunk < num_chunks[l_idx][j]; ++chunk) {
                     filenames.push_back(tmp_file(l_idx, j, chunk));
                 }
+
+                // if there are too many chunks, merge them into larger ones
+                while (filenames.size() > files_open_per_thread) {
```

Comment:
I would prefer all this merge logic to be in `common::merge_files` rather than here. Perhaps `merge_files` can take an extra parameter, `max_open_files` which limits the total number of files it's allowed to open when merging. Or just copy this section of code (module removing the first file) into a new method `merge_files_bounded`.

Reply:
yeah, I thought about that, but then thought we'd also need to pass a path for temp files, so it would change the interfaces there and looks like a bigger refactoring.